### ISSUE TITLE
Modify persistent connections

### ISSUE THREAD

**Issue by jinxing3114** [mongodb/mongo-php-driver#593]
@jmikola 
i try disable the hash creation
![image](https://cloud.githubusercontent.com/assets/13598919/25981238/919d977c-3706-11e7-9125-19cf033c10f9.png)
But the number of connections or growth. I would like to ask, how to modify the long connection to short connection

```
insert query update delete getmore command % dirty % used flushes vsize   res qr|qw ar|aw netIn netOut conn                      time
    *0    *0     *0     *0       0     1|0     0.0    1.1       0  281M  103M   0|0   0|0   79b  21.2k    3 2017-05-12T10:42:55+08:00
    *0    *0     *0     *0       0     1|0     0.0    1.1       0  281M  108M   0|0   0|0   79b  21.2k    3 2017-05-12T10:42:56+08:00
```

mongo-php-driver version 1.2.9

**Comment by jmikola** [comment]
If you are disabling the call to `php_phongo_manager_make_client_hash()`, you should also remove the calls to `php_phongo_find_client()` and `php_phongo_persist_client()` later in the function. In particular, `php_phongo_persist_client()` is going to result in persistent clients accumulating over the duration of the PHP worker process' lifetime. It's possible by leaving the `hash` empty you're overwriting existing clients in the call to `php_phongo_persist_client()`, which may result in the overwritten client being destroyed since we have `php_phongo_pclient_destroy()` configured as the destructor for hash elements; however, I can't be sure of that and it'd be safer to disable all of the aforementioned functions.

In a previous reply to https://github.com/mongodb/mongo-php-driver/issues/564#issuecomment-296087949, I described how to retain client persistence (useful if a single request constructs multiple Managers with the same URI) while ensuring that connections get destroyed at the end of each request. Quoting here:

> If you'd like to experiment with disabling connection persistence, you could try moving the hash initialization and destruction from the extensions' [`GINIT`](https://github.com/mongodb/mongo-php-driver/blob/master/php_phongo.c#L2045) and [`MSHUTDOWN`](https://github.com/mongodb/mongo-php-driver/blob/master/php_phongo.c#L2185) functions in `php_phongo.c`to [`RINIT`](https://github.com/mongodb/mongo-php-driver/blob/master/php_phongo.c#L2018) and [`RSHUTDOWN`](https://github.com/mongodb/mongo-php-driver/blob/master/php_phongo.c#L2198).
>
> I've linked the `master` branch above, but porting this to `v1.2` would require creating the `RINIT` and `RSHUTDOWN` functions and replacing the following lines in the `zend_module_entry mongodb_module_entry` declaration to register those functions:
>
> ```diff
> -	NULL /* PHP_RINIT(mongodb)*/,
> -	NULL /* PHP_RSHUTDOWN(mongodb)*/,
> +	PHP_RINIT(mongodb),
> +	PHP_RSHUTDOWN(mongodb),
> ```
>
> This change would ensure any sockets are closed after each request. New requests would start with an empty topology, which would make them ignorant of any sockets closed by your firewall configuration.

**Comment by jinxing3114** [comment]
mongo-php-driver version 1.2.9 modify
![image](https://cloud.githubusercontent.com/assets/13598919/26038924/b034ae78-3945-11e7-9c92-bf7da8df1f76.png)
![image](https://cloud.githubusercontent.com/assets/13598919/26038925/bcef1fb8-3945-11e7-8060-27268c52e3d3.png)
![image](https://cloud.githubusercontent.com/assets/13598919/26038928/c8427f22-3945-11e7-8de4-c15c23b668be.png)
but run make command failed

In file included from /usr/include/php/20151012/Zend/zend.h:36:0,
                 from /usr/include/php/20151012/main/php.h:36,
                 from /root/mongodb-1.2.9/php_phongo.c:27:
/root/mongodb-1.2.9/php_phongo.c: In function ‘zm_activate_mongodb’:
/usr/include/php/20151012/Zend/zend_API.h:173:76: error: ‘zend_mongodb_globals {aka struct _zend_mongodb_globals}’ has no member named ‘subscribers’
 #define ZEND_MODULE_GLOBALS_ACCESSOR(module_name, v) (module_name##_globals.v)
                                                                            ^
/usr/include/php/20151012/Zend/zend_hash.h:61:97: note: in definition of macro ‘zend_hash_init’
 #define zend_hash_init(ht, nSize, pHashFunction, pDestructor, persistent)      _zend_hash_init((ht), (nSize), (pDestructor), (persistent) ZEND_FILE_LINE_CC)
                                                                                                 ^
/root/mongodb-1.2.9/php_phongo.h:56:23: note: in expansion of macro ‘ZEND_MODULE_GLOBALS_ACCESSOR’
 # define MONGODB_G(v) ZEND_MODULE_GLOBALS_ACCESSOR(mongodb, v)
                       ^
/root/mongodb-1.2.9/php_phongo.c:1873:18: note: in expansion of macro ‘MONGODB_G’
  zend_hash_init(&MONGODB_G(subscribers), 0, NULL, ZVAL_PTR_DTOR, 0);
                  ^
In file included from /usr/include/php/20151012/main/php.h:40:0,
                 from /root/mongodb-1.2.9/php_phongo.c:27:
/root/mongodb-1.2.9/php_phongo.c: In function ‘zm_deactivate_mongodb’:
/usr/include/php/20151012/Zend/zend_API.h:173:76: error: ‘zend_mongodb_globals {aka struct _zend_mongodb_globals}’ has no member named ‘subscribers’
 #define ZEND_MODULE_GLOBALS_ACCESSOR(module_name, v) (module_name##_globals.v)
                                                                            ^
/root/mongodb-1.2.9/php_phongo.h:56:23: note: in expansion of macro ‘ZEND_MODULE_GLOBALS_ACCESSOR’
 # define MONGODB_G(v) ZEND_MODULE_GLOBALS_ACCESSOR(mongodb, v)
                       ^
/root/mongodb-1.2.9/php_phongo.c:2042:21: note: in expansion of macro ‘MONGODB_G’
  zend_hash_destroy(&MONGODB_G(subscribers));
                     ^
Makefile:529: recipe for target 'php_phongo.lo' failed
-------------------------------------------------
mongo-php-driver version master
run make command failed
In file included from /root/mongo-php-driver-master/src/libmongoc/src/mongoc/mongoc-read-prefs.h:27:0,
                 from /root/mongo-php-driver-master/src/libmongoc/src/mongoc/mongoc-server-description.h:23,
                 from /root/mongo-php-driver-master/src/libmongoc/src/mongoc/mongoc-apm.h:28,
                 from /root/mongo-php-driver-master/src/libmongoc/src/mongoc/mongoc-apm-private.h:25,
                 from /root/mongo-php-driver-master/src/libmongoc/src/mongoc/mongoc-apm.c:17:
/root/mongo-php-driver-master/src/libmongoc/src/mongoc/mongoc-config.h:156:30: error: operator '!=' has no left operand
 #if MONGOC_ENABLE_SASL_CYRUS != 1
                              ^
/root/mongo-php-driver-master/src/libmongoc/src/mongoc/mongoc-config.h:167:29: error: operator '!=' has no left operand
 #if MONGOC_ENABLE_SASL_SSPI != 1
                             ^
Makefile:320: recipe for target 'src/libmongoc/src/mongoc/mongoc-apm.lo' failed
make: *** [src/libmongoc/src/mongoc/mongoc-apm.lo] Error 1

please help me see what the problem is
thank you

**Comment by jmikola** [comment]
The build errors you shared with `MONGODB_G(subscribers)` and `MONGOC_ENABLE_SASL_CYRUS` should not be possible with the 1.2.9 tag or v1.2 branch. Those artifacts relate to code in the master branch.

My intention in the previous comment (and linked issue) was not to suggest copying RINIT and RSHUTDOWN code from the master branch as-is. I was only pointing out that we now define those handlers in the master branch for unrelated reasons. Altering v1.2's persistence behavior to clear `MONGODB_G(pclients)` after each request would require implementing those handlers.

If you're not familiar with PHP's extension API, I suggest sticking with my first suggestion. Starting with the 1.2.9 tag and corresponding git submodules (for libbson and libmongoc), follow the steps for [building from source|https://github.com/mongodb/mongo-php-driver/blob/master/CONTRIBUTING.md#building-from-source]. From there, read through `phongo_manager_init()` and experiment with disabling the logic around the following function calls:

 * `php_phongo_manager_make_client_hash()`
 * `php_phongo_find_client()`
 * `php_phongo_persist_client()`
